modules/react/combobox/lib/hooks/useKeyboardTypeAhead.ts

```diff
@@ -0,0 +1,132 @@
+ * This hook is best used when composing hooks and in conjunction with `useComboboxInput`.
+ *
+ * ```tsx
+ * // Example for a `Select` input
+ * export const useSelectInput = composeHooks(
+ * createElemPropsHook(useComboboxModel)((model, ref, elemProps: {keySofar?: string} = {}) => {
+ *  return {
+ *    onKeyDown(event: React.KeyboardEvent) {
+ *     // onKeyDown logic ...
+ *    },
+ *  } as const;
+ * }),
+ * useKeyboardTypeAhead,
+ * useComboboxInput
+ * );
+ * ```
+ */
+export const useKeyboardTypeAhead = createElemPropsHook(useComboboxModel)(model => {
+  const keySofar = React.useRef('');
+  const timer = React.useRef<ReturnType<typeof setTimeout>>();
+  const [keyTypedSofar, setKeyTypedSofar] = React.useState(keySofar.current);
+
+  const startClearKeysSoFarTimer = () => {
+    if (timer.current) {
+      clearTimeout(timer.current);
+    }
+    timer.current = setTimeout(() => {
+      keySofar.current = '';
+      setKeyTypedSofar(keySofar.current);
+    }, 500);
+  };
+
+  // Based on the the key typed, this will try and find an index in the array where it matches the string
+  // so if a user types `de` it will try and find an index in the array whose id matches `de` like denver
+  const getIndexByStartString = (
+    startIndex: number,
+    startString: string,
+    endIndex: number = model.state.items.length,
+    ignoreDisabled: boolean = true
+  ): number => {
+    for (let i = startIndex; i < endIndex; i++) {
+      const label = model.state.items[i].textValue.toLowerCase();
+      if (label.indexOf(startString.toLowerCase()) === 0) {
+        if (
+          !ignoreDisabled ||
+          (ignoreDisabled && !model.state.nonInteractiveIds.includes(model.state.items[i].id))
+        ) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  };
+
+  const currentItem = model.navigation.getItem(model.state.cursorId, model);
+  const cursorFocusedIndex = currentItem.index;
+
+  const handleKeyboardTypeAhead = (key: string, numOptions: number) => {
+    // If the starting point is beyond the list of options, reset it
+    // to the beginning of the list
+    const startNumber = keySofar.current.length === 0 ? cursorFocusedIndex + 1 : cursorFocusedIndex;
+
+    const start = startNumber === numOptions ? 0 : startNumber;
+
+    // Keeps track of the current key types and adds to it
+    // if you type `de` vs `d` for denver
+    keySofar.current += key;
+    setKeyTypedSofar(keySofar.current);
+    startClearKeysSoFarTimer();
+
+    // First, look for a match from start to end
+    let matchIndex;
+    matchIndex = getIndexByStartString(start, keySofar.current);
+
+    // If a match isn't found between start and end, wrap the search
+    // around and search again from the beginning (0) to start
+    if (matchIndex === -1) {
+      matchIndex = getIndexByStartString(0, keySofar.current, start);
+    }
```

Comment:
So after this step `matchIndex` will be always more than `-1` or there is any possibility `matchIndex` will be `0`?

Reply:
it's possible it's 0

Reply:
if it's the first item

Reply:
I think @RayRedGoose means to say "is there any possibility `matchIndex` will be `-1`"? A `0` means match found and it is the first item. This logic forces a wrap around from the current cursor, but it is possible a match is still not found.